Add spawner interface to allow selection of spawn points

arturkamieniecki opened the pull request:
I've added a spawnerInterace which finds all entities with the ROS2 spawn point component. This allows the display of all possible spawn points for the URDF importer. I've also added a list of available spawn points in the URDF importer and modified urdfPrefabMaker to spawn the imported robot in the selected place. If no spawn point is assigned a default one will be used.

michalpelka commented on Gems/ROS2/Code/CMakeLists.txt:
```diff
@@ -38,6 +38,7 @@ ly_add_target(
         PUBLIC
             AZ::AzCore
             AZ::AzFramework
+            AZ::AzToolsFramework
```
Sorry, but it is not acceptable to use AZ::AzToolsFramework in game mode. 
It makes the game executable with this Gem activated almost as heavy as Editor.

michalpelka reviewed (changes requested):
I cannot accept change, since it uses Editor only libraries in the Game code.
I would recommend to :
 - Create ROS2SpawnerEditorComponent that will be active in editor.
- Add ROS2SpawnerEditorInterfaceBus
- Call AZ::EBusAggregateResults<AZ::EntityId> .
Please refer to   `void EditorViewportWidget::SetSelectedCamera()` and  `Camera::CameraRequests::GetCameras` since that is almost identical case.

lumberyard-employee-dm commented on Gems/ROS2/Code/Source/Spawner/ROS2SpawnerInterface.cpp:
```diff
@@ -0,0 +1,96 @@
+#include <AzCore/RTTI/TypeInfoSimple.h>
+#include <AzCore/std/smart_ptr/shared_ptr.h>
+#include <AzCore/std/utils.h>
+#include <AzToolsFramework/API/ToolsApplicationAPI.h>
+#include <Entity/EditorEntityInfoBus.h>
+
+namespace ROS2
+{
+    const AZ::Transform& ROS2SpawnerInterface::GetDefaultSpawnPose() const
+    {
+        return m_defaultSpawnPose;
+    };
+
+    const AZStd::vector<AZStd::pair<AZStd::string, AZStd::shared_ptr<AZ::Transform>>> ROS2SpawnerInterface::GetAllSpawnPoints() const
+    {
+        AZStd::vector<AZStd::pair<AZStd::string, AZStd::shared_ptr<AZ::Transform>>> allSpawnPoints;
+
+        AZ::EntityId levelEntityId;
+        AzToolsFramework::ToolsApplicationRequestBus::BroadcastResult(
+            levelEntityId, &AzToolsFramework::ToolsApplicationRequests::GetCurrentLevelEntityId);
```
It looks like you want access to gather all the entities from the level that is loaded.

The ToolsApplication is not needed for that case.

The EntityContextEventBBus [OnEntityContextLoadedFromStream](https://github.com/o3de/o3de/blob/development/Code/Framework/AzFramework/AzFramework/Entity/EntityContextBus.h#L174-L179) notification EBus can be handled to get the list of game entities.

lumberyard-employee-dm replied:
Alternatively, you can

lumberyard-employee-dm commented on Gems/ROS2/Code/Source/Spawner/ROS2SpawnerInterface.h:
```diff
@@ -0,0 +1,34 @@
+
+        AZ::Transform m_defaultSpawnPose = { AZ::Vector3{ 0, 0, 0 }, AZ::Quaternion{ 0, 0, 0, 1 }, 1.0 };
+    };
+} // namespace ROS2
```
extraordinary nitpick: file is missing newline at the end.

lumberyard-employee-dm commented on Gems/ROS2/Code/Source/Spawner/ROS2SpawnerInterface.h:
```diff
@@ -0,0 +1,34 @@
+
+        // void registerNewSpawnPointComponent(const AZ::EntityId& entityId) override;
+
+        // void removeSpawnPointComponent(const AZ::EntityId& entityId) override;
+        const AZStd::vector<AZStd::pair<AZStd::string, AZStd::shared_ptr<AZ::Transform>>> GetAllSpawnPoints() const;
```
The return type shouldn't be a const value since a new instance is returned.
I also I recommend making a type alias for Spawn vector to help document what the `vector<pair<string, shared_ptr<AZ::Transform>>>` is for.
```suggestion
        using SpawnPoint = AZStd::pair<AZStd::string, AZStd::shared_ptr<AZ::Transform>>;
        using SpawnPointVector = AZStd::vector<SpawnPoint>;
        SpawnPointVector GetAllSpawnPoints() const;
```

lumberyard-employee-dm commented on Gems/ROS2/Code/Source/Spawner/ROS2SpawnerInterface.cpp:
```diff
@@ -0,0 +1,96 @@
+        }
+    }
+
+} // namespace ROS2
```
This file is missing a new line.

lumberyard-employee-dm commented on Gems/ROS2/Code/Source/Spawner/ROS2SpawnerInterface.cpp:
```diff
@@ -0,0 +1,96 @@
+            AZ::ComponentApplicationBus::BroadcastResult(entity, &AZ::ComponentApplicationBus::Events::FindEntity, entityId);
+            AZ_Assert(entity, "No child entity %s", entityId.ToString().c_str());
+
+            const auto allComponents = entity->GetComponents();
+
+            for (const AZ::Component* component : allComponents)
+            {
+                if (component->GetUnderlyingComponentType() == ROS2SpawnPointTypeId)
+                {
+                    auto name = entity->GetName();
+                    allSpawnPoints.push_back(
+                        AZStd::make_pair(name, AZStd::make_shared<AZ::Transform>(entity->GetTransform()->GetWorldTM())));
+                }
+            }
```
This should use `entity->FindComponent(ROS2SpawnPointTypeId)` to locate entities that has a ROS2SpawnPoint Component on it.

lumberyard-employee-dm commented on Gems/ROS2/Code/Include/ROS2/Spawner/SpawnerInterface.h:
```diff
@@ -0,0 +1,34 @@
+    };
+
+    using SpawnerInterface = AZ::Interface<ISpawner>;
+} // namespace ROS2
```
nitpick: missing newline at the end of the file.

lumberyard-employee-dm commented on Gems/ROS2/Code/Include/ROS2/Spawner/SpawnerInterface.h:
```diff
@@ -0,0 +1,34 @@
+        //! scale: 1.0)
+        virtual const AZ::Transform& GetDefaultSpawnPose() const = 0;
+
+        //! All available spawn pose getter
+        //! @return vector of all available spawn positions from all entities having a ROS2 spawn point component.
+        //! This vector consists of pairs with the entity name and its pose.
+        virtual const AZStd::vector<AZStd::pair<AZStd::string, AZStd::shared_ptr<AZ::Transform>>> GetAllSpawnPoints() const = 0;
```
This comment https://github.com/o3de/o3de-extras/pull/343/files#r1200718835 about not returning const for a vector being returned by value should be applied to this file.

arturkamieniecki commented:
I've applied the suggestion from @michalpelka. The new code differs a lot from the original. There are many many files changed, but most of these changes are re-writing the existing components. I've modified the spawner and spawn point components to use a controller and a configuration. I've also added Editor components for both of them. 
The spawner editor component handles EBus requests for gathering information about available spawn points. 
The spawner game component also handles these requests but in the game mode.
I've also modified the FleetRobotTemplate warehouse.prefab to add the new components. The components configuration remains the same, only the component type was changed.

adamdbrw commented:
@michalpelka @adamdbrw for re-review

michalpelka commented on Gems/ROS2/Code/Source/RobotImporter/Pages/PrefabMakerPage.cpp:
```diff
@@ -60,5 +95,24 @@ namespace ROS2
     {
         return m_useArticulation->isChecked();
     }
+    AZStd::shared_ptr<AZ::Transform> PrefabMakerPage::getSelectedSpawnPoint() const
```
Why shared pointer is used here? you can use optional or return zero matrix.

michalpelka reviewed (approved):
A lot of work good work here!
I do not like UX 
![image](https://github.com/o3de/o3de-extras/assets/3209244/c296b23c-dc79-4ece-964b-61578a7ea484)
but it should not be a blocker. I think we should polish it in next PR.
Minor thing: this PR is not compiling without rebase. Please rebase it before AR.
Nitpicks in the submitted code.

michalpelka commented:
It was a major revision. It would be very beneficial for us to have @lumberyard-employee-dm take a second look.

byrcolin reviewed (changes requested):
looks mostly good

arturkamieniecki commented:
@byrcolin could you take another look at the applied changes?

adamdbrw commented:
@byrcolin could you take another look if the changes you requested have been implemented to your satisfaction?

spham-amzn commented on Gems/ROS2/Code/Source/RobotImporter/Pages/PrefabMakerPage.cpp:
```diff
@@ -17,6 +28,20 @@ namespace ROS2
+        m_spawnPointsComboBox = new QComboBox(this);
+        m_spawnPointsInfos = allActiveSpawnPoints.values;
+
+        for (auto i = 0; i < allActiveSpawnPoints.values.size(); i++)
```
`auto` is unnecessary here. 
```suggestion
        for (int i = 0; i < allActiveSpawnPoints.values.size(); i++)
```

spham-amzn commented on Gems/ROS2/Code/Source/RobotImporter/Pages/PrefabMakerPage.cpp:
```diff
@@ -28,6 +53,17 @@ namespace ROS2
+        QLabel* spawnPointListLabel;
+        if (allActiveSpawnPoints.values.size() == 0)
+        {
+            spawnPointListLabel = new QLabel("Select spawn position (No spawn position were detected)", this);
```
```suggestion
            spawnPointListLabel = new QLabel("Select spawn position (No spawn positions were detected)", this);
```

spham-amzn commented on Gems/ROS2/Code/Source/RobotImporter/Pages/PrefabMakerPage.cpp:
```diff
@@ -60,5 +96,24 @@ namespace ROS2
+        if (!m_spawnPointsInfos.empty())
+        {
+            int vectorIndex = m_spawnPointsComboBox->currentData().toInt();
+            AZStd::string mapKey = AZStd::string(m_spawnPointsComboBox->currentText().toStdString().c_str());
```
This invokes both a constructor and a move operator, why not just initialize it directly?
```suggestion
            AZStd::string mapKey(m_spawnPointsComboBox->currentText().toStdString().c_str());
```

spham-amzn commented on Gems/ROS2/Code/Source/RobotImporter/Pages/PrefabMakerPage.cpp:
```diff
@@ -60,5 +96,24 @@ namespace ROS2
+            int vectorIndex = m_spawnPointsComboBox->currentData().toInt();
+            AZStd::string mapKey = AZStd::string(m_spawnPointsComboBox->currentText().toStdString().c_str());
+            auto& map = m_spawnPointsInfos[vectorIndex];
+            auto spawnInfo = map.find(mapKey);
+            if (spawnInfo == m_spawnPointsInfos[vectorIndex].end())
+            {
+                return AZStd::nullopt;
+            }
+            return spawnInfo->second.pose;
+        }
+        else
+        {
+            return AZStd::nullopt;
+        }
+    }
```
This section can be simplified
```suggestion
            if (auto spawnInfo = map.find(mapKey);
                spawnInfo != map..end())
            {
                return spawnInfo->second.pose;
            }
        }
        return AZStd::nullopt;
    }
```

spham-amzn commented on Gems/ROS2/Code/Source/RobotImporter/URDF/URDFPrefabMaker.cpp:
```diff
@@ -323,13 +325,13 @@ namespace ROS2
+        if (!spawnPosition.has_value())
         {
-            AZ_TracePrintf("URDF Importer", "Spawner not found - creating entity in (0,0,0)\n") return;
+            AZ_TracePrintf("URDF Importer", "SpawnPosition is null - spawning in Editors default position\n");
```
`AZ_TracePrintf` is deprecated. Use `AZ_Trace` instead
```suggestion
            AZ_Trace("URDF Importer", "SpawnPosition is null - spawning in Editors default position\n");
```

spham-amzn commented on Gems/ROS2/Code/Source/RobotImporter/URDF/URDFPrefabMaker.cpp:
```diff
@@ -340,9 +342,8 @@ namespace ROS2
-
-        transformInterface_->SetWorldTM(pose);
+        transformInterface_->SetWorldTM(*spawnPosition);
+        AZ_TracePrintf("URDF Importer", "Successfully set spawn position\n")
```
```suggestion
        AZ_Trace("URDF Importer", "Successfully set spawn position\n")
```

spham-amzn commented on Gems/ROS2/Code/Source/RobotImporter/URDF/URDFPrefabMaker.cpp:
```diff
@@ -340,9 +342,8 @@ namespace ROS2
             AZ_TracePrintf("URDF Importer", "TransformComponent not found in created entity\n") return;
```
While you are editing this file (and others) can you replace all `AZ_TracePrintf` with `AZ_Trace`?

spham-amzn commented on Gems/ROS2/Code/Source/Spawner/ROS2SpawnerComponentController.cpp:
```diff
@@ -0,0 +1,135 @@
+        {
+            AZ::Entity* childEntity = nullptr;
+            AZ::ComponentApplicationBus::BroadcastResult(childEntity, &AZ::ComponentApplicationRequests::FindEntity, child);
+            AZ_Assert(childEntity, "No child entity %s", child.ToString().c_str());
```
```suggestion
            AZ_Assert(childEntity, "No child entity found for entity %s", child.ToString().c_str());
```

spham-amzn commented on Gems/ROS2/Code/Source/Spawner/ROS2SpawnerComponentController.cpp:
```diff
@@ -0,0 +1,135 @@
+            AZ::ComponentApplicationBus::BroadcastResult(childEntity, &AZ::ComponentApplicationRequests::FindEntity, child);
+            AZ_Assert(childEntity, "No child entity %s", child.ToString().c_str());
+
+            const auto* spawnPoint = childEntity->FindComponent<ROS2SpawnPointComponent>();
+
+            if (spawnPoint == nullptr)
+            {
+                continue;
+            }
+
+            result.insert(spawnPoint->GetInfo());
```
```suggestion
            if (const auto* spawnPoint = childEntity->FindComponent<ROS2SpawnPointComponent>();
               spawnPoint != nullptr)
            {
                result.insert(spawnPoint->GetInfo());
            }
```

spham-amzn commented on Gems/ROS2/Code/Source/Spawner/ROS2SpawnerComponentController.h:
```diff
@@ -0,0 +1,65 @@
+        //////////////////////////////////////////////////////////////////////////
+        // SpawnerRequestsBus::Handler overrides
+        const AZ::Transform& GetDefaultSpawnPose() const override;
+        AZStd::unordered_map<AZStd::string, SpawnPointInfo> GetAllSpawnPointInfos() const override;
```
I noticed this in SpawnerBus.h:
```
using SpawnPointInfoMap = AZStd::unordered_map<AZStd::string, SpawnPointInfo>;
```
We can probably use it for this and all other references to `AZStd::unordered_map<AZStd::string, SpawnPointInfo>`

```suggestion
        SpawnPointInfoMap GetAllSpawnPointInfos() const override;
```

spham-amzn commented on Gems/ROS2/Code/Source/Spawner/ROS2SpawnerComponentController.h:
```diff
@@ -0,0 +1,65 @@
+        //////////////////////////////////////////////////////////////////////////
+
+        AZStd::unordered_map<AZStd::string, SpawnPointInfo> GetSpawnPoints() const;
+        ROS2SpawnerComponentConfig m_config;
```
should make this private
```suggestion
    private:
        ROS2SpawnerComponentConfig m_config;
```

spham-amzn commented on Gems/ROS2/Code/Source/Spawner/ROS2SpawnerEditorComponent.cpp:
```diff
@@ -0,0 +1,102 @@
+        {
+            AZ::Entity* childEntity = nullptr;
+            AZ::ComponentApplicationBus::BroadcastResult(childEntity, &AZ::ComponentApplicationRequests::FindEntity, child);
+            AZ_Assert(childEntity, "No child entity %s", child.ToString().c_str());
```
```suggestion
            AZ_Assert(childEntity, "No child entity found for entity %s", child.ToString().c_str());
```

spham-amzn commented on Gems/ROS2/Code/Source/Spawner/ROS2SpawnerEditorComponent.cpp:
```diff
@@ -0,0 +1,102 @@
+    {
+        ROS2SpawnerEditorComponentBase::Activate();
+        ROS2SpawnerComponentConfig config = m_controller.GetConfiguration();
+        config.m_editorEntityId = GetEntityId();
```
We may want to add an assert here to make sure `GetEntityId()` doesn't return null/invalid entityid.

spham-amzn commented on Gems/ROS2/Code/Source/Spawner/ROS2SpawnerEditorComponent.cpp:
```diff
@@ -0,0 +1,102 @@
+        ROS2SpawnerComponentConfig config = m_controller.GetConfiguration();
+        config.m_editorEntityId = GetEntityId();
+        m_controller.SetConfiguration(config);
+        SpawnerRequestsBus::Handler::BusConnect(GetEntityId());
```
No need to call `GetEntityId()` twice
```suggestion
        SpawnerRequestsBus::Handler::BusConnect(config.m_editorEntityId);
```

spham-amzn commented on Gems/ROS2/Code/Source/Spawner/ROS2SpawnerEditorComponent.cpp:
```diff
@@ -0,0 +1,102 @@
+
+    void ROS2SpawnerEditorComponent::Deactivate()
+    {
+        ROS2SpawnerEditorComponentBase::Deactivate();
+        SpawnerRequestsBus::Handler::BusDisconnect();
```
In general, we should perform the disconnect in the reverse order of connect in `Activate`
```suggestion
        SpawnerRequestsBus::Handler::BusDisconnect();
        ROS2SpawnerEditorComponentBase::Deactivate();
```

arturkamieniecki commented:
Applied suggestions and resolved conflicts.

michalpelka commented:
@byrcolin, kindly ask for the review to unblock merging. Thanks!
